numpy2 compatibility

This workflow has been failing for more than 2 weeks and I had not realized because I did not get any notifications.

https://github.com/nilearn/nilearn/actions/workflows/numpy2_compatibility.yml

Failing tests: https://github.com/nilearn/nilearn/actions/runs/7665540512/job/20891619867#step:5:18881

```
FAILED nilearn/regions/tests/test_region_extractor.py::test_remove_small_regions
FAILED nilearn/regions/tests/test_region_extractor.py::test_connected_label_regions
FAILED nilearn/regions/tests/test_region_extractor.py::test_connected_label_regions_return_empty_for_large_min_size
FAILED nilearn/regions/tests/test_region_extractor.py::test_connected_label_regions_check_labels
```

To run locally:

```bash
tox run -e test_numpy2 -- nilearn/regions/tests/test_region_extractor.py
```

Also realising that numpy 2 is actually not installed by this workflow:

```
test_numpy2: freeze> python -m pip freeze --all
test_numpy2: certifi==2023.11.17,charset-normalizer==3.3.2,contourpy==1.2.0,coverage==7.4.0,cycler==0.12.1,fonttools==4.47.2,idna==3.6,iniconfig==2.0.0,joblib==1.3.2,kaleido==0.2.1,kiwisolver==1.4.5,lxml==5.1.0,matplotlib==3.8.2,nibabel==5.2.0,nilearn @ file:///home/runner/work/nilearn/nilearn/.tox/.tmp/package/1/nilearn-0.1.dev1%2Bgd0cbdcd.tar.gz#sha256=694501acd0e789448bcc3a688a97f8ccbf04f8895f5e684780597fc1a8d83d31,numpy==1.26.3,packaging==23.2,pandas==2.2.0,pillow==10.2.0,pip==23.3.1,plotly==5.18.0,pluggy==1.4.0,pyparsing==3.1.1,pytest==7.4.4,pytest-cov==4.1.0,python-dateutil==2.8.2,pytz==2023.3.post1,requests==2.31.0,scikit-learn==1.4.0,scipy==1.12.0,six==1.16.0,tenacity==8.2.3,threadpoolctl==3.2.0,tzdata==2023.4,urllib3==2.1.0
```


Potentially useful:

- numpy 2 migration guide: https://numpy.org/devdocs/numpy_2_0_migration_guide.html
- [numpy/numpy#24300](https://github.com/numpy/numpy/issues/24300)
- scikit learn issue about it: https://github.com/scikit-learn/scikit-learn/issues/27075
- https://pythonspeed.com/articles/numpy-2/

--- a/.github/workflows/numpy2_compatibility.yml
+++ b/.github/workflows/numpy2_compatibility.yml
@@ -4,10 +4,6 @@
 name: Numpy2 compatibility
 
 on:
-    # for testing
-    pull_request:
-        branches:
-        -   '*'
     schedule:
     # Run every day at 8am UTC
     -   cron: 0 8 * * *

--- a/.github/workflows/numpy2_compatibility.yml
+++ b/.github/workflows/numpy2_compatibility.yml
@@ -4,6 +4,10 @@
 name: Numpy2 compatibility
 
 on:
+    # for testing
+    pull_request:
+        branches:
+        -   '*'
     schedule:
     # Run every day at 8am UTC
     -   cron: 0 8 * * *
@@ -20,6 +24,11 @@ jobs:
             uses: actions/setup-python@v5
             with:
                 python-version: 3.12
+        -   name: install and run ruff checks for NPY NPY201
+            # see https://numpy.org/devdocs/numpy_2_0_migration_guide.html#ruff-plugin
+            run: |
+                python -m pip install ruff>0.1.8
+                ruff check --preview --select NPY201 .
         -   name: Install tox
             run: python -m pip install tox
         -   name: Run test suite

--- a/nilearn/regions/region_extractor.py
+++ b/nilearn/regions/region_extractor.py
@@ -116,7 +116,7 @@ def _remove_small_regions(input_data, index, affine, min_size):
     # np.unique and then use np.bincount to count the region sizes.
 
     _, region_indices = np.unique(input_data, return_inverse=True)
-    region_sizes = np.bincount(region_indices)
+    region_sizes = np.bincount(region_indices.ravel())
     size_in_vox = min_size / np.abs(np.linalg.det(affine[:3, :3]))
     labels_kept = region_sizes > size_in_vox
     if not np.all(labels_kept):

--- a/tox.ini
+++ b/tox.ini
@@ -150,6 +150,8 @@ setenv =
 extras = test
 deps =
     {[plotting]deps}
+; recreating the environment to avoid dependency conflict when not starting from a clean slate
+recreate = true
 commands =
     pip install git+https://github.com/nipy/nibabel
     pip install --pre --upgrade -i {env:nightlies_url} pandas scipy scikit-learn matplotlib

